File: app/src/main/java/com/rud/movieapp/view/base/MvvmActivity.java
package com.rud.movieapp.view.base;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.databinding.library.baseAdapters.BR;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;

import javax.inject.Inject;

public class MvvmActivity<B extends ViewDataBinding, V extends ViewModel & LifecycleObserver> extends AppCompatActivity {

    protected B binding;
    protected V viewModel;
    @Inject
    Class<V> viewModelClass;
    @Inject
    ViewModelProvider.Factory viewModelFactory;

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    protected final View setAndBindContentView(@NonNull AppCompatActivity activity, @Nullable Bundle savedInstanceState, @LayoutRes int layoutResID) {
        setContentView(layoutResID);
        binding = DataBindingUtil.setContentView(this, layoutResID);
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(viewModelClass);
        binding.setVariable(BR.vm, viewModel);
        activity.getLifecycle().addObserver(viewModel);
        return binding.getRoot();
    }
}

File: app/src/main/java/com/rud/movieapp/view/base/MvvmActivity.kt
package com.rud.movieapp.view.base

import android.os.Bundle
import android.view.View
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.databinding.library.baseAdapters.BR
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import javax.inject.Inject

open class MvvmActivity<B : ViewDataBinding, V> :
    AppCompatActivity() where V : ViewModel, V : LifecycleObserver {
    @Inject
    lateinit var viewModelClass: Class<V>
    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory
    private lateinit var binding: B
    lateinit var viewModel: V

    public override fun onDestroy() {
        super.onDestroy()
    }

    protected fun setAndBindContentView(
        activity: AppCompatActivity,
        savedInstanceState: Bundle?, @LayoutRes layoutResID: Int
    ): View {
        setContentView(layoutResID)
        binding = DataBindingUtil.setContentView(this, layoutResID)
        viewModel = ViewModelProviders.of(this, viewModelFactory)[viewModelClass]
        binding.setVariable(BR.vm, viewModel)
        activity.lifecycle.addObserver(viewModel)
        return binding.root
    }
}